### PR TITLE
Update air pipeline for conv2d lowering, using conv-decompose pipeline

### PR DESCRIPTION
**Pull request by erwei-xilinx**
- Requires https://github.com/Xilinx/mlir-air/pull/633 to land in AIR.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp [hunk 1]
```diff
@@ -529,9 +529,14 @@ void buildAMDAIETransformPassPipeline(OpPassManager &variantPassManager) {
...
-    addMLIRAIRLoweringPasses(modulePassManager, false);
+    addMLIRAIRLoweringPasses(modulePassManager,
+                             AIEPassPipeline::PadPackPipeline);
   } else if (clUsePipeline == AIEPassPipeline::PackPeelPipeline) {
-    addMLIRAIRLoweringPasses(modulePassManager, true);
+    addMLIRAIRLoweringPasses(modulePassManager,
+                             AIEPassPipeline::PackPeelPipeline);
+  } else if (clUsePipeline == AIEPassPipeline::ConvDecomposePipeline) {
+    addMLIRAIRLoweringPasses(modulePassManager,
+                             AIEPassPipeline::ConvDecomposePipeline);
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:539]: I don't think you need to pass AIEPassPipeline as a variable. Instead, you can call `addMLIRAIRLoweringPasses(modulePassManager)` directly and use `clUsePipeline` inside the function directly.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp [hunk 2]
```diff
@@ -546,13 +551,14 @@ void buildAMDAIETransformPassPipeline(OpPassManager &variantPassManager) {
...
 // for details.
-void addMLIRAIRLoweringPasses(OpPassManager &passManager, bool packPeel) {
+void addMLIRAIRLoweringPasses(OpPassManager &passManager,
+                              AIEPassPipeline passPipeline) {
   // Add passes for preparing for lowering to MLIR-AIR
   passManager.addPass(createEraseHALDescriptorTypeFromMemRefPass());
   passManager.addPass(memref::createFoldMemRefAliasOpsPass());
   passManager.addPass(createAMDAIEBridgeToAIRPass());
   // TODO (Erwei): Figure out a way to work with AMDAIEPackToDmaPass.
-  if (packPeel)
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:555]: If this flag is not used anymore, we should also clean up the flag at the top of this file.

**Review by yzhang93 — CHANGES_REQUESTED** [review]
Thanks for pushing the Conv2d pipeline in MLIR-AIR! LGTM mostly, some comments to address.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp [hunk 3]
```diff
@@ -668,18 +676,22 @@ void addMLIRAIRLoweringPasses(OpPassManager &passManager, bool packPeel) {
...
     xilinx::air::AffineLoopOptPassOptions options;
     std::vector<unsigned> tile_sizes;
-    if (packPeel) {
+    if (passPipeline == AIEPassPipeline::PackPeelPipeline) {
       tile_sizes = {2, 2};
-    } else {
+    } else if (passPipeline == AIEPassPipeline::PadPackPipeline) {
       tile_sizes = {4, 4};
-    }
+    } else
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:686]: Please add a comment in the code what this code is for, and make sure that when in the future the set of passes in AIEPassPipeline changes, it is easy to understand (assert?)

**Comment by yzhang93** [comment]
@newling After this PR merged, would you help add ci tests for Conv2d similar to what we have for matmul? I think we can use what is used in IREE https://github.com/iree-org/iree/blob/main/tests/e2e/convolution/generate_e2e_conv2d_tests.py.

**Comment by newling** [comment]
> @newling After this PR merged, would you help add ci tests for Conv2d similar to what we have for matmul? I think we can use what is used in IREE https://github.com/iree-org/iree/blob/main/tests/e2e/convolution/generate_e2e_conv2d_tests.py.

You can easily add a test in cpu_comparison, @erwei-xilinx  could add it in this PR if there is a numerically passing test?

**Review by newling — COMMENTED** [review]
Thanks, good to see it working with AIR! Happy to help with testing (@yzhang93)

**Comment by erwei-xilinx** [comment]
> > @newling After this PR merged, would you help add ci tests for Conv2d similar to what we have for matmul? I think we can use what is used in IREE https://github.com/iree-org/iree/blob/main/tests/e2e/convolution/generate_e2e_conv2d_tests.py.
> 
> You can easily add a test in cpu_comparison, @erwei-xilinx could add it in this PR if there is a numerically passing test?

Yeah would be really useful if we can add the numerical test in this PR.

**Comment by yzhang93** [comment]
> > @newling After this PR merged, would you help add ci tests for Conv2d similar to what we have for matmul? I think we can use what is used in IREE https://github.com/iree-org/iree/blob/main/tests/e2e/convolution/generate_e2e_conv2d_tests.py.
> 
> You can easily add a test in cpu_comparison, @erwei-xilinx could add it in this PR if there is a numerically passing test?

Yes, we can have a quick numerical test using cpu_comparison at the moment. But in the long term, I think we'd want to use IREE's stand alone test module. I think it's faster and more flexible?

**Comment by newling** [comment]
Test can be added something like this: https://github.com/newling/iree-amd-aie/tree/add_conv_test

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp (hunk 1 above)
Comment (erwei-xilinx) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp]: Make sense. Updated.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp (hunk 2 above)
Comment (erwei-xilinx) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:558]: Although this variable packPeel is no longer used, the AIR pipeline for pad-pack, pack-peel and conv-decompose still differ in some places, which were implemented with some if conditions.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp (hunk 3 above)
Comment (erwei-xilinx) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:678]: Make sense. Comments added.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp [hunk 4]
```diff
@@ -531,11 +531,7 @@ void buildAMDAIETransformPassPipeline(OpPassManager &variantPassManager) {
         [&]() { return createAMDAIELowerExecutableTargetPass(options); });
   }
   modulePassManager.addPass(createLowerUKernelOpsToCallsPass());
-  if (clUsePipeline == AIEPassPipeline::PadPackPipeline) {
-    addMLIRAIRLoweringPasses(modulePassManager, false);
-  } else if (clUsePipeline == AIEPassPipeline::PackPeelPipeline) {
-    addMLIRAIRLoweringPasses(modulePassManager, true);
-  }
+  addMLIRAIRLoweringPasses(modulePassManager, clUsePipeline);
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:534]: I think you can simplify this function further. You don't have to pass `clUsePipeline` to this function.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp [hunk 5]
```diff
@@ -581,7 +578,8 @@ void addMLIRAIRLoweringPasses(OpPassManager &passManager, bool packPeel) {
   passManager.addPass(createCSEPass());
 
   passManager.addPass(xilinx::air::createAIRDependencyPass());
-  if (!(packPeel && clMatmulElementwiseFusion)) {
+  if (!(passPipeline == AIEPassPipeline::PackPeelPipeline &&
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:581]: Replace `passPipeline` with `clUsePipeline`.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp (hunk 2 above)
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:558]: NVM... I thought there was a separate flag, but looks like there's not.

**Comment by erwei-xilinx** [comment]
> Test can be added something like this: https://github.com/newling/iree-amd-aie/tree/add_conv_test

I cherry picked this commit and the test is passing. May I cherr-pick your commit into this PR and we land together?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp (hunk 4 above)
Comment (erwei-xilinx) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp]: Make sense. Updated.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp (hunk 5 above)
Comment (erwei-xilinx) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp]: Updated.

**Review by yzhang93 — APPROVED** [review]

**Comment by newling** [comment]
> > Test can be added something like this: https://github.com/newling/iree-amd-aie/tree/add_conv_test
> 
> I cherry picked this commit and the test is passing. May I cherr-pick your commit into this PR and we land together?

Wow that's surprising that it just worked. Thank you @erwei-xilinx !

**Comment by erwei-xilinx** [comment]
> > > Test can be added something like this: https://github.com/newling/iree-amd-aie/tree/add_conv_test
> > 
> > 
> > I cherry picked this commit and the test is passing. May I cherr-pick your commit into this PR and we land together?
> 
> Wow that's surprising that it just worked. Thank you @erwei-xilinx !

Thanks for putting together the test!

**Pull request merged**